src/pudl/transform/eia.py

```diff
@@ -1149,133 +1098,188 @@ def fix_balancing_authority_codes_with_state(
+            for name, df in clean_dfs.items()
+        }
+
+        if entity == EiaEntity.UTILITIES:
+            # Remove location columns that are associated with plants, not utilities:
+            for table, df in clean_dfs.items():
+                if "plant_id_eia" in df.columns:
+                    plant_location_cols = [
+                        "street_address",
+                        "city",
+                        "state",
+                        "zip_code",
+                    ]
+                    logger.info(f"Removing {plant_location_cols} from {table} table.")
+                    clean_dfs[table] = df.drop(
+                        columns=plant_location_cols, errors="ignore"
+                    )
 
-    Args:
-        eia_transformed_dfs (dict): a dictionary of table names (keys) and
-            transformed dataframes (values).
+        entity_df, annual_df, col_dfs = harvesting(
+            entity, clean_dfs, debug=debug, eia860m=eia_settings.eia860.eia860m
+        )
 
-    Returns:
-        Dataframes of entity tables transformed EIA tables.
-    """
-    # Do some final cleanup and assign appropriate types:
-    eia_transformed_dfs = {
-        name: convert_cols_dtypes(df, data_source="eia")
-        for name, df in eia_transformed_dfs.items()
-    }
+        # Apply standard PUDL data types to the new entity tables:
+        pkg = Package.from_resource_ids()
+        entity_res = pkg.get_resource(f"{entity.value}_entity_eia")
+        entity_df = apply_pudl_dtypes(entity_df, group="eia").pipe(entity_res.encode)
+        annual_res = pkg.get_resource(f"{entity.value}_eia860")
+        annual_df = apply_pudl_dtypes(annual_df, group="eia").pipe(annual_res.encode)
+
+        if entity == EiaEntity.PLANTS:
+            # Post-processing specific to the plants entity tables
+            entity_df = _add_additional_epacems_plants(entity_df).pipe(_add_timezone)
+            annual_df = fillna_balancing_authority_codes_via_names(annual_df).pipe(
+                fix_balancing_authority_codes_with_state, plants_entity=entity_df
+            )
 
-    # Remove the clean_ prefix from the table names.
-    eia_transformed_dfs = {
-        table_name.replace("clean_", ""): df
-        for table_name, df in eia_transformed_dfs.items()
-    }
+        entity_df = entity_res.enforce_schema(entity_df)
+        annual_df = annual_res.enforce_schema(annual_df)
+        # Need to figure out how to return the col_dfs if debug is True
+        return (
+            Output(output_name=f"{entity.value}_entity_eia", value=entity_df),
+            Output(output_name=f"{entity.value}_eia860", value=annual_df),
+        )
 
-    eia_settings = context.resources.dataset_settings.eia
+    return harvested_entity
 
-    # create the empty entities df to fill up
-    entities_dfs = {}
 
-    # for each of the entities, harvest the static and annual columns.
-    # the order of the entities matter! the
+def debug_harvesting(
```

Comment:
Is there really no simple way to avoid this duplication? It does simplify the notebook considerably though, and the function should also be getting tested independently by virtue of the notebook integration tests.

Reply:
Could we just call `debug_harvesting()` in the `harvested_entity` multi_asset?

```python
def harvested_entity_asset_factory(
    entity: EiaEntity, io_manager_key: str | None = None
) -> AssetsDefinition:
    """Create an asset definition for the harvested entity tables."""
    harvestable_assets = (
        "clean_boiler_fuel_eia923",
        ...
    )

    @multi_asset(...)
    def harvested_entity(context, **clean_dfs):
        """Harvesting IDs & consistent static attributes for EIA entity."""
        logger.info(f"Harvesting IDs & consistent static attributes for EIA {entity}")
        eia_settings = context.resources.dataset_settings.eia
        debug = context.op_config["debug"]
        # Do some final cleanup and assign appropriate types:
        clean_dfs = {
            name: convert_cols_dtypes(df, data_source="eia")
            for name, df in clean_dfs.items()
        }
        entity_df, annual_df, col_dfs = debug_harvesting(entity, clean_dfs, eia_settings=eia_settings) # Should probably rename this function
        return (
            Output(output_name=f"{entity.value}_entity_eia", value=entity_df),
            Output(output_name=f"{entity.value}_eia860", value=annual_df),
        )
```

Reply:
Or maybe it would be simpler to just roll all of the additional logic that is surrounding the call to `harvesting()` in the asset factory into `harvesting()` itself, since we need to run it every time we run `harvesting()` anyway! 🤦🏼

Reply:
That sounds good to me! As long as `harvest()` doesn't use any dagster objects, we can call it in a notebook.

Reply:
I knew I was making this too complicated.